lld/lld-c/LLDAsLibraryC.cpp

```diff
@@ -732,68 +665,312 @@ resolveEVMLinkerSymbols(MemoryBufferRef inBuffer,
-                                   char ***linkerSymbols,
-                                   uint64_t *numLinkerSymbols) {
-  if (linkerSymbols) {
-    *numLinkerSymbols = 0;
-    *linkerSymbols = nullptr;
+/// Create a linker script to produce 'assembly' ELF object file.
+/// An example of script:
+///
+///   ENTRY(0);
+///   SECTIONS {
+///     . = 0;
+///     .text : SUBALIGN(1) {
+///       __datasize__$12c297efd8baf$__ = 268;
+///       __dataoffset__$b5e66d52578d$__ = ABSOLUTE(.);
+///       __$b5e66d52578d$__(.text);
+///       __dataoffset__$12c297efd8baf$__ = ABSOLUTE(.);
+///       __$12c297efd8baf$__(.text);
+///       __datasize__$b5e66d52578d$__ = ABSOLUTE(.);
+///     }
+///     /DISCARD/ : {
+///     }
+///  }
+///  VERSION {
+///  { local:
+///    __datasize__$12c297efd8baf$__;
+///    __dataoffset__$b5e66d52578d$__;
+///    __dataoffset__$12c297efd8baf$__;
+///    __datasize__$b5e66d52578d$__;
+///  };
+///  };
+
+static std::string createEVMAssembeScript(ArrayRef<LLVMMemoryBufferRef> memBufs,
+                                          ArrayRef<const char *> depIDs,
+                                          const StringSet<> &depsToLink) {
+  assert(memBufs.size() == depIDs.size());
+
+  // Maps symbol name section to an array of object IDs that contain it.
+  StringMap<SmallVector<StringRef>> symbolNameSectionsMap;
+
+  auto getDataSizeName = [](StringRef name) {
+    return EVM::getDataSizeSymbol(EVM::getLinkerSymbolHash(name));
+  };
+
+  // A set of all the defined symbols within the script that should
+  // have later be removed from the symbol table. For this, we set
+  // their visibility to 'local'. These symbols will later be removed via
+  // 'objcopy' API.
+  std::string definedSymbolsBuf;
+  raw_string_ostream definedSymbols(definedSymbolsBuf);
+
+  std::unique_ptr<Binary> topLevelBinary =
+      cantFail(createBinary(unwrap(memBufs[0])->getMemBufferRef()));
+  const auto *topLevelObjFile =
+      static_cast<const ObjectFile *>(topLevelBinary.get());
+  getSymbolNameSections(topLevelObjFile, depIDs[0], symbolNameSectionsMap);
+
+  std::string textSectionBuf;
+  raw_string_ostream textSection(textSectionBuf);
+
+  // Define symbols whose values represent sizes of the dependencies.
+  // Skip the topLevelObject, as its size denotes .text section size of
+  // the resulting object file.
+  for (unsigned idx = 1; idx < memBufs.size(); ++idx) {
+    std::unique_ptr<Binary> binary =
+        cantFail(createBinary(unwrap(memBufs[idx])->getMemBufferRef()));
+    const auto *objFile = static_cast<const ObjectFile *>(binary.get());
+    std::string bufIdHash = EVM::getLinkerSymbolHash(depIDs[idx]);
+    if (depsToLink.count(bufIdHash))
+      getSymbolNameSections(objFile, depIDs[idx], symbolNameSectionsMap);
+
+    for (const SectionRef &sec : objFile->sections()) {
+      if (!sec.isText())
+        continue;
+
+      StringRef secName = cantFail(sec.getName());
+      if (secName == ".text") {
```

Comment:
Can we have multiple text section with different names? Otherwise can we assert?

Reply:
Nope, we should have only text section, so I added assertion instead of `if`